osu.Game.Rulesets.Osu/OsuRuleset.cs

```diff
@@ -328,5 +328,21 @@ public override StatisticItem[] CreateStatisticsForScore(ScoreInfo score, IBeatm
         }
 
         public override RulesetSetupSection CreateEditorSetupSection() => new OsuSetupSection();
+
+        public double PreemptFromAr(float AR) => AR < 5 ? (1200.0 + 600.0 * (5 - AR) / 5) : (1200.0 - 750.0 * (AR - 5) / 5);
+        public float ArFromPreempt(double preempt) => (float)(preempt > 1200 ? ((1800 - preempt) / 120) : ((1200 - preempt) / 150 + 5));
+        public double HitwindowFromOd(float OD) => 80.0 - 6 * OD;
+        public float OdFromHitwindow(double hitwindow300) => (float)(80.0 - hitwindow300) / 6;
+        public float ChangeArFromRate(float AR, double rate) => ArFromPreempt(PreemptFromAr(AR) / rate);
+        public float ChangeOdFromRate(float OD, double rate) => OdFromHitwindow(HitwindowFromOd(OD) / rate);
```

Comment:
What on earth is this mess..

Reply:
I wanted to put all these functions that are usually just copied to different places into one place.
I shouldn't do this? I can put all of this code inside the function `GetRateAdjustedDifficulty`

Reply:
> I wanted to put all these functions that are usually just copied to different places into one place.

You may have identified other usages, but you haven't updated any of them to use these new functions. Ignoring the fact that these functions have no xmldoc and are named badly, it doesn't make sense to do this unless you also centralise all existing usages.

If you're going to do this, please do it in a separate PR as a prerequisite or follow-up to this change. For now I'd recommend inlining these.